### ISSUE TITLE
Analysis results paths are pointing to generated code instead of the Razor files

### ISSUE THREAD

**Issue by costin-zaharia-sonarsource** [dotnet/roslyn#71449]
For the diagnostics reported in `.razor` files, the locations are pointing to the generated files instead of the actual razor files.

E.g. when a diagnostic is reported on the `IncrementCount` identifier from the following code (`Counter.razor` file)
```
<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
```
The Sarif report contains:
```
          "locations": [
            {
              "resultFile": {
                "uri": "Microsoft.NET.Sdk.Razor.SourceGenerators/Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator/Components_Pages_Counter_razor.g.cs",
```
You can find [here](https://github.com/dotnet/razor/files/13615158/BlazorWebApp.zip) a reproducer. It contains already the generated Sarif report (`BlazorWebApp\BuildOutput.sarif`)

When working on custom code analyzers and development tools in general, it is very useful to have access to the original location of the syntax. For [sonar-dotnet](https://github.com/SonarSource/sonar-dotnet) analyzers, we need the correct locations to be able to import and display the issues on SonarQube and SonarCloud.

We tried to do the mapping ourselves (by using [GetMappedLineSpan](https://learn.microsoft.com/es-es/dotnet/api/microsoft.codeanalysis.syntaxtree.getmappedlinespan?view=roslyn-dotnet-4.6.0)) but the API offers limited functionality, and we often encounter issues like https://github.com/dotnet/roslyn/issues/69248. This has also unexpected behavior in the IDE like https://github.com/dotnet/razor/issues/9308 or https://github.com/dotnet/razor/issues/9108.

@davidwengier explained [here](https://github.com/dotnet/razor/issues/9308#issuecomment-1730846267) that:
> Also, I'm not surprised line mappings don't work, as they are not very sophisticated without the Razor compiler moving to enhanced line pragmas. I would expect that if a diagnostic is reported on the unmapped location, it probably has a better chance of working, as we will correctly map it to the .razor file, and we have more source mapping information than the Roslyn APIs can see. I don't think Roslyn's internal analyzers generally call GetMappedSpan() when creating diagnostics.

# tl;dr

Would it be possible to change paths for the diagnostics in razor files to point to the original files instead of the generated ones?

**Comment by RikkiGibson** [comment]
Resolved in #71454
